pkg/agent/openflow/pipeline.go

```diff
@@ -142,9 +167,27 @@ var (
 	// snatMarkRange takes the 17th bit of register marksReg to indicate if the packet needs to be SNATed with Node's IP
 	// or not. Its value is 0x1 if yes.
 	snatMarkRange = binding.Range{17, 17}
+	// hairpinMarkRange takes the 18th bit of register marksReg to indicate
+	// if the packet needs DNAT to virtual IP or not. Its value is 0x1 if yes.
+	hairpinMarkRange = binding.Range{18, 18}
+	// macRewriteMarkRange takes the 19th bit of register marksReg to indicate
+	// if the packet needs to be rewrite MAC address. Its value is 0x1 if yes.
+	macRewriteMarkRange = binding.Range{19, 19}
+	// endpointIPRegRange takes a 32-bit range of register endpointIPReg to store
+	// the selected Service Endpoint IP.
+	endpointIPRegRange = binding.Range{0, 31}
+	// endpointPortRegRange takes a 16-bit range of register endpointPortReg to store
+	// the selected Service Endpoint port.
+	endpointPortRegRange = binding.Range{0, 15}
+	// serviceLearnRegRange takes a 3 bits range of register serviceLearnReg to
+	// indicate if the Service accessing packet is selected the Service Endpoint,
+	// needs to select an Endpoint or needs to be learned the Endpoint selection
+	// decision.
```

Comment:
I don't understand this sentence. Do you mean:
> serviceLearnRegRange takes a 3 bits range of register serviceLearnReg to indicate if the packet accessing a Service has already selected the Service Endpoint, still needs to select an Endpoint, or if an Endpoint has already been selected and the selection decision needs to be learned.

Reply:
Thanks, this is what I want to express. Updated.